sysfs numa distances as costs

ffromani opened the pull request:
We need to fill the NUMA zone cost information.
We start using NUMA distances, as reported by linux sysfs, as costs.

Test PR to check the waters, intended to be used mostly as conversation starter to evaluate the approach.

Once we get agreement about the important topics I will either polish up this PR or file a cleaner new one.

There are a number of design decisions to discuss:
1. are we happy with minimal code to read directly from sysfs?
   do we want to introduce/depend on an external package?
   please not cadvisor doesn't seem to report NUMA distances/costs
   (yet?)
2. Who do we make sure we reconcile properly the NUMA zones from
   podresources to sysfs?
3. (minor, but still) why do topologyapi and the internal protocol
   use slightly different types (e.g int vs int32)? can we get
   uniformity?
4. do we want a deeper refactoring of `pkg/finder`?

ffromani commented on pkg/topologyinfo/sysfs/sysfs.go:
```diff
@@ -0,0 +1,79 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/fromanirh/cpuset"
```
this is wrong, we need to use kube's cpuset package here. This change is pretty much transparent so I'll leave it for now.

swatisehgal commented on cmd/nfd-topology-updater/main.go:
```diff
@@ -136,7 +134,7 @@ func argsParse(argv []string) (topology.Args,finder.Args, error) {
   --key-file=<path>           Private key matching --cert-file
                               [Default: ]
-  --server=<server>           NFD server address to connecto to.
+  --server=<server>           NFD server address to connect to.
```
Thanks for catching these typos!

swatisehgal commented on pkg/nfd-topology-updater/dumpobject.go:
```diff
@@ -0,0 +1,29 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
This might need to be updated

ffromani replied:
fixed by https://github.com/swatisehgal/node-feature-discovery/pull/5/ and rebased accordingly!

swatisehgal commented on pkg/nfd-master/nfd-master.go:
```diff
@@ -576,39 +572,38 @@ func addAnnotations(n *api.Node, annotations map[string]string) {
+		zones[zoneName] = v1alpha1.Zone{
+			Type:   zone.Type,
+			Parent: zone.Parent,
+			// Costs: updateMap(zone.Costs),
```
The zone cost is not being updated here which means that the CRD is not created with the cost. Just uncommenting this would make it work.

ffromani replied:
right! let me fix this

swatisehgal reviewed (changes requested):
> 1. are we happy with minimal code to read directly from sysfs?

As a starting point, I am happy with the way this PR evaluates the cost, we can certainly get feedback from others (within the team and externally) to explore other viable options.
>    do we want to introduce/depend on an external package?
>    please not cadvisor doesn't seem to report NUMA distances/costs

I came across [this](https://github.com/google/cadvisor/issues/1521) which clearly indicates the intension to enhance cadvisor to expose topology information. From my reading, the topology information was eventually exposed but it only includes Nodes, cores, threads, per-node memory, and caches but doesn't seem to expose distances/costs that we would like to be exposed. So, sticking to the current approach of using sysfs seems fine to me. Alternatively, we can create an issue in cadvisor and see if this is something that would be useful for the community.
> 2. Who do we make sure we reconcile properly the NUMA zones from
>    podresources to sysfs?

Can you elaborate what you mean by this?
> 3. (minor, but still) why do topologyapi and the internal protocol
>    use slightly different types (e.g int vs int32)? can we get
>    uniformity?

This discrepancy is because proto file does not allow us to use `int` hence the use of int32 in topology-updater.proto. We can certainly modify it in the topologyapi but felt a bit odd to make changes to an API to adapt to the code constraints of the NFD! 

> 4. do we want a deeper refactoring of `pkg/finder`?

I am happy with the creation of topologyinfo package which we are now using in finder/noderesource.go. In my opinion, now the code looks much more streamlined and neat  but I am always up for improvement so refactor away if you like! I am here to review :)

ffromani committed:
master: propagate NUMA zone costs

We need to pass the costs through to make them reach
the CRDs.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
pkg: topo-updater: add  and use the "dumpobject" helper

In order to reduce the number of dependencies,
introduce a simple object dump function based on `encoding/json`,
and use it to replace the `spew` package.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
topology: use sysfs NUMA distances as cost

Add code to read NUMA distances from sysfs, and use these values
as Costs for NUMA zones.

There are a number of design decisions to discuss:
1. are we happy with minimal code to read directly from sysfs?
   do we want to introduce/depend on an external package?
   please not cadvisor doesn't seem to report NUMA distances/costs
   (yet?)
2. Who do we make sure we reconcile properly the NUMA zones from
   podresources to sysfs?
3. (minor, but still) why do topologyapi and the internal protocol
   use slightly different types (e.g int vs int32)? can we get
   uniformity?
4. do we want a deeper refactoring of `pkg/finder`?

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
topologyinfo: sysfs: use k8s cpuset package

We already depend on kubelet, so we can and should use
the kubernetes' cpuset package.

Signed-off-by: Francesco Romani <fromani@redhat.com>
